pkg/controllers/management/authprovisioningv2/role.go

```diff
@@ -153,6 +157,83 @@ func (h *handler) OnChange(key string, rt *v3.RoleTemplate) (*v3.RoleTemplate, e
 	return rt, nil
 }
 
+func (h *handler) hasClusterAnnotationsAndDeletionTimestamp(annotations map[string]string) (bool, error) {
+	clusterName, cOK := annotations[clusterNameLabel]
+	clusterNamespace, cnOK := annotations[clusterNamespaceLabel]
+	if !cOK || !cnOK {
+		// if this role doesn't have a clustername/namespace label it isn't one of the protected roles, so move on
+		return false, nil
+	}
+	cluster, err := h.clusters.Get(clusterNamespace, clusterName)
```

Comment:
This function currently appears to be using the controllers for the v2 provisioning clusters. There are two different cluster objects - `clusters.provisioning.cattle.io` and `clusters.management.cattle.io`. V1 makes heavier use of the second type of cluster while v2 provisioning makes heavier use of the first type of cluster. The second type, or management clusters, are not namespaced resources, which requires cluster roles/ cluster role bindings for users to access them. 

The code that you have here will interact well with provisioning clusters but will not interact well with v1 clusters, which will cause some issues with RKE1 clusters. I would recommend re-evaluating your approach to do the following:

1. Include a cache object in the handler for the management type clusters
2. Configure the `onRemoveClusterRole` and `onRemoveClusterRoleBinding` methods to use this object to make sure that the cluster still exists. There's a couple of ways that you can do this - either by passing in a "DoesClusterExist" function as a parameter which you construct, using a boolean flag to indicate if the check should be done for mgmt or provisioning, or by just checking both clusters and making sure that neither of them could be found.

Reply:
@MbolotSuse What are the benefits of checking both cluster CRDs over just the cluster CRD presented to the UI?

Reply:
Both clusters are presented to the UI, just in different places. Since they both are presented, you need to check both of them. 

This is also leaving aside automation (such as creation of v1 prov vs v2 prov clusters) which would likely need access to the "source of truth" CRD.